Update to Twemoji 15.1.0

TheEssem opened the pull request:
Fixes #32700. This introduces the emojis from the [Twemoji v15.1.0](https://github.com/jdecked/twemoji/releases/tag/v15.1.0) release, conforming to the Emoji 15.1 standard. It includes support in both the renderer and emoji picker. 🐦‍🔥🍄‍🟫

Since `@emoji-mart/data` still doesn't support Emoji 15.1 and the `emoji-datasource` package hasn't been updated to include the Twemoji 15.1.0 graphics, some Rake scripts have been added to generate both the spritesheet (using libvips) and the emoji data JSON (using the same compressed format as emoji-mart v3 so little further conversion is needed, generated with data from https://github.com/iamcal/emoji-data and https://github.com/muan/emojilib) locally.

TheEssem committed:
Update to Twemoji 15.1.0

TheEssem committed:
Remove `@emoji-mart/data`

It's only being used for the category names now, which can be moved into the repo directly since they're so small.

TheEssem committed:
Fix prettierignore

TheEssem commented:
Seems like the failing tests are due to some keywords added as a result of an emojilib update I applied locally in the emoji-mart data generation. Will fix.

TheEssem committed:
Fix keywords in emoji_all

renchap commented:
Thanks a lot for your PR!

I dont really like adding files that are based on open PRs on other repos, as it will make it harder to know how to regenerate them in the future.

A better solution would probably to copy the script that generate those files into our repo and simplify to only keep what is relevant to us, so we no longer depend on those 3rd party repositories for future updates, and the process to generate the files is something under our control. Do you think this can be done without too much work?

TheEssem committed:
Add Rake script to generate emoji spritesheet

TheEssem commented:
That should be doable, yeah. I've started by adding a Rake script (`emojis:generate_emoji_sheet`) to generate the spritesheet based on the local emoji sheet JSON and SVG files.

Marking as draft until the JSON files can be generated as well.

renchap commented:
If it is easier, we can have those scripts written in JS rather than Ruby, as they require JS packages (and we can probably copy the emoji-mart one as a base?)

TheEssem commented:
> If it is easier, we can have those scripts written in JS rather than Ruby, as they require JS packages (and we can probably copy the emoji-mart one as a base?)

The JS packages are simply wrappers for JSON data (with the exception of `inflection` which is just a port of Rails inflections), so there wouldn't be much advantage to using JS over Ruby/Rake for this.

I feel like it would probably be a good idea to consolidate the different emoji JSON files into one or two, since they're starting to get a bit unwieldy with a fair bit of duplicate data...

TheEssem committed:
Generate unified emoji data JSON

The Rake script generates data in the emoji-mart pre-v5 compressed format, so the previous changes to convert a v5 format data file are no longer needed. This also means that emoji_all.json and emoji_categories.json are no longer needed, and since the generated file contains accurate sheet coordinates, emoji_sheet.json is no longer needed either.

TheEssem committed:
Fix failing test

TheEssem commented:
It appears that Twemoji has now been updated to v16: https://github.com/jdecked/twemoji/releases/tag/v16.0.1

It should be possible to use the Rake scripts to generate the new emoji picker data/spritesheet required for a major emoji version update provided that emoji-data and emojilib have updated as well. Unfortunately, however, neither seem to have updated yet.

ClearlyClaire commented:
Looks like an improvement overall but I'm bit surprised by the replacement of the specific family emoji variants with very generic white-silhouette-on-blue-background icons. Is that a bug? Or an Twemoji deliberate change?

renchap commented:
Thanks a lot!

We probably need to go a step further and generate `emoji_compressed.json` directly from your Rake task (https://github.com/mastodon/mastodon/pull/34554 does it in JS), but this can be done in a further PR.
